### ISSUE TITLE
Unable to Turn On Audio Description.

### ISSUE THREAD

**Issue by kwizrr** [Sandmann79/xbmc#271]
Audio Descriptions narrate information about actions, characters, scene changes, on-screen text, and other visual content.Audio Descriptions narrate information about actions, characters, scene changes, on-screen text, and other visual content.

Turn Audio Descriptions On or Off
1.  Start playback of a movie or TV show with Audio Descriptions available.
2.  Select the Subtitles and Audio option in your playback controls.
3.  Select the audio track with the [Audio Description] tag or the desired audio language without the tag.

You can find a complete list of compatible Titles of Movies and TV Shows with Audio Description here:
https://www.amazon.com/video/audiodescriptions

Is there an easy way to allow individuals needing this very important accessibility feature to enable it (Perminantly or temporarily) in the add-on? Currently running Amazon VOD v0.6.9 in Kodi v18.1 Leia under Windows 10 v1809. Thanks for any help.

**Comment by Varstahl** [comment]
At the moment I don't see anything specific to audio descriptions in the files metadata. But I see, amongst other things, that for The Grand Tour there are 2 separate stereo tracks groups and 1 for dolby surround, so as reported by Amazon I'm guessing that one of them has audio descriptions.

I guess we could add an option to suppress audio tracks without audio description in them, so that the correct one gets picked by default, or something of the sort. I need to investigate this when I have time, I'll see what we can do about it.

**Comment by kwizrr** [comment]
Thanks. Such an option is very important for thousands of individuals like myself who rely on audio description to enjoy popular TV shows and movies.

**Comment by Varstahl** [comment]
With 9a0cac11ecd0d7e64cfe60c630987561c365544f I introduced a first support to Audio Description. If you go to the Addon Settings, Language, Audio, and enable Audio descriptions, when watching a movie or a show you should be able to select the audio description tracks.

This is a first step, but unfortunately at this time we don't have the ability to automatically select the Audio Description tracks. I will try to provide it later on, but in the meanwhile the update with these changes should arrive on your machine in a day or two.

**Comment by kwizrr** [comment]
This is great news! Thanks for adding this option to your already excellant add-on. Believe me, it's very much appreciated.

**Comment by kwizrr** [comment]
Ok, I just received the Amazon VOD update containing the Audio Description option in Settings -> Language. Enabled it to "Yes." Now I'm listening/watching a TV show episode in Kodi Leia v18.1 with an Audio Description track. How do I switch to it? Is there a keyboard shortcut to access this menu? Tried both [c] and [m], but it doesn't seem to bring up the necessary menu to make the switch.

Thanks again for your help.

**Comment by Varstahl** [comment]
The only keyboard shortcuts to switch audio tracks as far as I know, with the video playing in full screen, are control+a, or long press of the up arrow key.

For the graphical menu, you can select the settings from the full screen video control toolbar to open up the video settings. The last menu item will cycle through the audio tracks just like the keyboard shortcut, or you can select the first menu item "audio settings", then select the menu item "audio stream", and from that list you can pick whatever language you prefer.

**Comment by kwizrr** [comment]
Oh boy! Think they must of remove the keyboard shortcut [Ctrl+a] from Kodi 18 Leia for it doesn't work. The Kodi Screenreader isn't able to read graphics, so being able to access the audio menu that way wouldn't work at all.

Has to be some way to re-enable the ability to control the switching of audio tracks via the keyboard, while the video is in full screen mode. I mean I can press [y] to switch the video player if I want, can't the same functionality be made for switching audio tracks too?

Maybe creating a special keyboard.xml file to enable this keyboard shortcut while the Amazon VOD add-on is in the foreground or video is playing full screen in the foreground.

Without this keyboard shortcut, this very important feature wouldn't be accessible to users wanting to access Amazon Prime audio described content via the Amazon VOD add-on.

**Comment by kwizrr** [comment]
Ok, I've been trying to learn more about keyboard shortcuts in Kodi (Action IDs) and came across the following info at: https://kodi.wiki/index.php?title=Action_IDs

ID: AudioNextLanguage
Description: Change to the next audio track in a video with multiple audio tracks.

Think the above ID is what needs to be linked to a keyboard shortcut to select the audio described track?

Problem is I have no programming skills for creating a proper keyboard.xml file to do this. Can this be done from within Amazon VOD?

**Comment by kwizrr** [comment]
Well been playing around with this a little more on my own.

I manually added the following line to the "service.xbmc.tts.keyboard.xml" file, which the Kodi Screenreader created in the /Kodi/userdata/keymaps folder under Windows 10.

<f7>XBMC.NotifyAll(service.xbmc.tts,AudioNextLanguage)</f7>

So I then played an audio described episode from the Prime original series "Informer." Made sure the video was in full screen mode. Hit [f7] which is supposed to switch to the next audio track. It just keeps playing the original one for some odd reason. Maybe I'm doing something wrong here?

Again using Kodi v18.1 Leia.

**Comment by kwizrr** [comment]
Ok, got it working now. I had to change the Stream Selection option in the Input Stream dependency settings from "Manual" to "Auto." Also the [Ctrl+a] keyboard shortcut does work for cycling between the different audio language tracks. Sorry about my rambling. Just reset the keyboard file for my Kodi to default, and all seemed to work fine. Thanks for a great job.

**Comment by kwizrr** [comment]
Ok, it works. The [Ctrl+a] shortcut key does cycle between the audio language tracks. Works great! Thanks for a fantastic job.

**Comment by Varstahl** [comment]
Oh, no problem about "the "rambling", I had no idea it didn't work with manual selection either, I learnt something new today. I'm glad to hear it works fine, no pun intended.

**Comment by Essam3115** [comment]
@kwizrr If you are still using the add-on this workaround might help you in selecting the Audio Description by default, It worked for me with the English language, Enable Audio Descriptions in the settings (this workaround might disable the audio if Audio Descriptions not enabled), in `proxy.py` replace

                    trackId = re.search(r'\s+audioTrackId="([a-z]{2})(-[a-z0-9]{2,})_(dialog|descriptive)', setTag).groups()
                    lang = re.search(r'\s+lang="([a-z]{2})"', setTag).group(1)

with


                    trackId = re.search(r'\s+audioTrackId="([a-z]{2})(-[a-z0-9]{2,})_(dialog|descriptive)', setTag).groups()
                    lang = re.search(r'\s+lang="([a-z]{2})"', setTag).group(1)
                    if 'dialog' == trackId[2]:
                        continue


![18](https://user-images.githubusercontent.com/85903831/122317453-da653880-cf1d-11eb-9d66-ca8b06f8e066.png)

**Comment by Essam31115** [comment]
If you usually watch only movies and TV shows that have Audio Descriptions and want to select the audio track with Audio Descriptions by default, enable Audio Descriptions in the settings and in `network.py` replace

`audioTrackId=all`

with

`audioTrackId=en-us_descriptive_0`

and the audio tracks with Audio Descriptions will be the only audio tracks that are available.

![111](https://user-images.githubusercontent.com/86118997/122610994-dbfb4180-d080-11eb-942c-8d2a11cbfb53.png)

**Comment by Varstahl** [comment]
That only works for US, it would break international support. We're going to do something similar in the proxy automatically, but didn't find the time to implement it yet.

**Comment by Essam31115** [comment]
@Varstahl I tested it with primevideo and it worked (France), do you mean if the language of the audio descriptions audio is not English?

**Comment by Varstahl** [comment]
Yes, if we hardcode english then for all other properly audio descripted languages it won't work, for example italian. In the optimal world, the user would select a range of languages and preferred audio description, and then the language could be selected while loading the stream, so that the stream loads even faster.

The side effect of enabling standard AD is also the complete removal of DD+ 5.1 tracks. Video might be related to that.

I've been wanting to do this for the past year and a half, but left the feature incomplete because of a number of reasons, including COVID. It will come, but not until I can weasel out of my precarious condition, or some other developer picks up the custom proxy embedded in the app. Logically it's simple, but it's quite tricky to implement, I'm unsure of the feasibility yet.

**Comment by Essam31115** [comment]
@Varstahl I found a workaround to select only the subtitle in my language by modifying `proxy.by`, and modified inputstream.adaptive to enable only the audio in aac (2 channels audio), I watch only American TV series and replacing `audioTrackId=all` with `audioTrackId=en-us_dialog_0` selects only the English audio tracks but it works only when I enable Audio descriptions which lowers the resolution (I'm now sure that disabling the regular expression  `mpd = re.sub(r'(~|%7E)', '', mpd)` removes 540p video from the mpd file) is it possible to made a simple modification in the add-on so `audioTrackId=en-us_dialog_0` works even when I do not enable Audio descriptions in the settings? (I know that workarounds are not the best solutions but workarounds all I can do until a proper fix is done)
